### PR TITLE
Allow adding extra repository for custom dependencies

### PR DESCRIPTION
**Pull request by xhaakon**

### CODE REVIEW

File: script.sh [hunk 1]
```diff
@@ -177,12 +178,28 @@ fi
 
 cat >>Dockerfile <<EOF
 RUN apt-get update && apt-get dist-upgrade --yes
-RUN apt-get install --yes --no-install-recommends build-essential equivs devscripts git-buildpackage
+RUN apt-get install --yes --no-install-recommends build-essential equivs devscripts git-buildpackage apt-transport-https wget ca-certificates
```
Comment (lamby) [inline script.sh]: Can this be split out?

File: script.sh [hunk 2]
```diff
@@ -177,12 +178,28 @@ fi
...
```
Comment (lamby) [inline script.sh:229]: - Can this be split out into a different pull request?

File: script.sh [hunk 3]
```diff
@@ -177,12 +178,28 @@ fi
...
+RUN wget -O- "${TRAVIS_EXTRA_REPOSITORY_GPG_URL}" | apt-key add -
+EOF
+fi
+
+if [ -n "${TRAVIS_EXTRA_REPOSITORY}" ]
+then
+	cat >>Dockerfile <<EOF
+RUN echo "deb ${TRAVIS_EXTRA_REPOSITORY}" >> /etc/apt/sources.list
+RUN echo "deb-src ${TRAVIS_EXTRA_REPOSITORY}" >> /etc/apt/sources.list
+RUN apt-get update
```
Comment (lamby) [inline script.sh:248]: Do we not run this later?

File: script.sh (hunk 1 above)
Comment (xhaakon) [inline script.sh]: Sure, but please see below...

File: script.sh (hunk 3 above)
Comment (xhaakon) [inline script.sh:248]: I can't find any later invocation of `apt-get update`, only the one above on line 180. I'm doing it here again, because the extra repository may use HTTPS, in which case I need `apt-transport-https` installed before I attempt to download from it. The process goes as follows:
1. add standard Debian repositories
2. `apt-get update && apt-get dist-upgrade`
3. install some base tools, `apt-transport-https` among them
4. add extra repository
5. `apt-get update` again in order to get package list from extra repository

That's also why I see the change in `apt-get install` you commented on above as an integral part of this commit and would rather if we kept everything in one piece... unless you really want me to split it.

Comment (lamby) [inline script.sh:248]: > unless you really want me to split it.

Into some kind of "extra_packages" or similar? Not sure we should be just throwing apt-transport-https in there by default unless for good reason..?

Comment (xhaakon) [inline script.sh:248]: The installation can be made optional.

I can check if `TRAVIS_EXTRA_REPOSITORY` is set and has HTTPS transport, in which case I'll put `apt-transport-https` and `ca-certificates` into `extra_packages`. And if we have `TRAVIS_EXTRA_REPOSITORY_GPG_URL`, I'll add `wget` as well.

Let me update the commit...

Comment (lamby) [reply script.sh:248]: Please do.

Comment (xhaakon) [reply script.sh:248]: > Please do

Sorry, what do you want me to do? The changes I mentioned above I already implemented a while ago.

I only noticed that the PR needed a rebase, so I've made it up-to-date.

File: script.sh [hunk 4]
```diff
@@ -208,14 +209,40 @@ RUN echo "deb-src ${TRAVIS_DEBIAN_MIRROR} experimental main" >> /etc/apt/sources
 EOF
 fi
 
+if [ "$(echo ${TRAVIS_EXTRA_REPOSITORY} | cut -c1-6)" = "https:" ]
```
Comment (lamby) [inline script.sh]: Please use `case` here to do the pattern matching

File: script.sh (hunk 3 above)
Comment (lamby) [reply script.sh:248]: Could you add a comment here why we run `apt-get update again` ie for https, preferably changing the code that we only run it for that case.

File: script.sh [hunk 5]
```diff
@@ -149,6 +149,7 @@ Info "Using distribution: ${TRAVIS_DEBIAN_DISTRIBUTION}"
 Info "Backports enabled: ${TRAVIS_DEBIAN_BACKPORTS}"
 Info "Experimental enabled: ${TRAVIS_DEBIAN_EXPERIMENTAL}"
 Info "Security updates enabled: ${TRAVIS_DEBIAN_SECURITY_UPDATES}"
+Info "Will use extra repository: ${TRAVIS_EXTRA_REPOSITORY:-<not set>}"
```
Comment (lamby) [inline script.sh]: Missing output of `TRAVIS_EXTRA_REPOSITORY_GPG_URL` here

File: script.sh [hunk 6]
```diff
@@ -208,14 +209,40 @@ RUN echo "deb-src ${TRAVIS_DEBIAN_MIRROR} experimental main" >> /etc/apt/sources
 EOF
 fi
 
+if [ "$(echo ${TRAVIS_EXTRA_REPOSITORY} | cut -c1-6)" = "https:" ]
+then
+	EXTRA_PACKAGES="apt-transport-https"
+fi
+
+if [ -n "${TRAVIS_EXTRA_REPOSITORY_GPG_URL}" ]
```
Comment (lamby) [inline script.sh]: I'd also like to see some (brief!) documentation addition to `index.html` — no need for it to be perfect, it just has to quickly go through what these new variables do and their defaults.

**Review by lamby — CHANGES_REQUESTED** [review]
Looking great, please just tidy up these last few things and I'll happily merge :)

**Comment by xhaakon** [comment]
Please check the latest version, which fixes the findings from your review.

File: script.sh (hunk 5 above)
Comment (lamby) [inline script.sh]: Still missing ``TRAVIS_EXTRA_REPOSITORY_GPG_URL``

File: script.sh [hunk 7]
```diff
@@ -208,14 +210,45 @@ RUN echo "deb-src ${TRAVIS_DEBIAN_MIRROR} experimental main" >> /etc/apt/sources
...
+# We're adding the extra repository only after the essential tools have been
+# installed, so that we have apt-transport-https if the repository needs it.
+# The subsequent apt-get update also avoids useless re-downloading of Debian
+# package lists we already have.
+if [ -n "${TRAVIS_EXTRA_REPOSITORY}" ]
+then
+	cat >>Dockerfile <<EOF
+RUN echo "deb ${TRAVIS_EXTRA_REPOSITORY}" >> /etc/apt/sources.list.d/travis_extra.list
+RUN echo "deb-src ${TRAVIS_EXTRA_REPOSITORY}" >> /etc/apt/sources.list.d/travis_extra.list
+RUN apt-get update -o Dir::Etc::sourcelist=/etc/apt/sources.list.d/travis_extra.list -o Dir::Etc::sourceparts=- --no-list-cleanup
```
Comment (lamby) [inline script.sh]: Hang on, what's withall the `-o` stuff? That needs a comment.

File: index.html [hunk 8]
```diff
@@ -167,6 +167,22 @@ <h3 id="configuration">
         Which Debian mirror to use. By default, <tt>ftp.de.debian.org</tt> is used.
       </dd>
       <dt>
+        TRAVIS_EXTRA_REPOSITORY
+      </dt>
+      <dd>
+        If the package has some dependencies not (yet) found in Debian, you can
```
Comment (lamby) [inline index.html]: please drop "(yet)" - maybe they will never be in Debian *shrugs*

File: index.html [hunk 9]
```diff
@@ -167,6 +167,22 @@ <h3 id="configuration">
...
+        add your own repository to the build using this variable. The format is
+        the same as in <tt>/etc/apt/sources.list</tt>, for example
```
Comment (lamby) [inline index.html]: as in → as

File: index.html [hunk 10]
```diff
@@ -167,6 +167,22 @@ <h3 id="configuration">
...
+        <tt>"https://my.repo.net/debian/ stretch main"</tt>.
+      </dd>
+      <dt>
+        TRAVIS_EXTRA_REPOSITORY_GPG_URL
+      </dt>
+      <dd>
+        From where to download the public key of the extra repository, so that
```
Comment (lamby) [inline index.html]: "From where to download" → "Location of"

"extra repository" → "extra repository defined in TRAVIS_EXTRA_REPOSITORY"

File: index.html [hunk 11]
```diff
@@ -167,6 +167,22 @@ <h3 id="configuration">
         Which Debian mirror to use. By default, <tt>ftp.de.debian.org</tt> is used.
       </dd>
       <dt>
+        TRAVIS_EXTRA_REPOSITORY
```
Comment (lamby) [inline index.html]: Please rename this variable to `TRAVIS_DEBIAN_EXTRA_REPOSITORY`

**Review by lamby — CHANGES_REQUESTED** [review]

**Comment by lamby** [comment]
Changes requested!

File: index.html (hunk 11 above)
Comment (xhaakon) [reply index.html]: I assume you also want `TRAVIS_DEBIAN_EXTRA_REPOSITORY_GPG_URL`...

File: script.sh (hunk 7 above)
Comment (xhaakon) [reply script.sh]: Those parameters make `apt-get update` act only on the just added extra repository, so that we avoid useless re-downloading of the Debian package lists we already have. I've extended the script comment to include what each of the arguments is doing.

**Comment by xhaakon** [comment]
Please see the latest revision.

File: script.sh [hunk 12]
```diff
@@ -208,14 +210,51 @@ RUN echo "deb-src ${TRAVIS_DEBIAN_MIRROR} experimental main" >> /etc/apt/sources
 EOF
 fi
```
Comment (lamby) [inline script.sh:212]: Missing a `EXTRA_PACKAGES = ""` I think, otherwise we break under `set -u`

File: script.sh [hunk 13]
```diff
@@ -208,14 +210,51 @@ RUN echo "deb-src ${TRAVIS_DEBIAN_MIRROR} experimental main" >> /etc/apt/sources
...
+then
+	cat >>Dockerfile <<EOF
+RUN wget -O- "${TRAVIS_DEBIAN_EXTRA_REPOSITORY_GPG_URL}" | apt-key add -
+EOF
+fi
+
+# We're adding the extra repository only after the essential tools have been
+# installed, so that we have apt-transport-https if the repository needs it.
+#
+# The subsequent apt-get update syncs only the content of the extra repository,
```
Comment (lamby) [inline script.sh]: Do we really need this? It seems unncessarily complicated (and unfamiliar to most users) that it doesn't warrant the complexity...

**Review by lamby — CHANGES_REQUESTED** [review]

File: script.sh (hunk 12 above)
Comment (xhaakon) [reply script.sh:212]: Yes we should add this.

File: script.sh (hunk 13 above)
Comment (xhaakon) [reply script.sh]: Saves a couple of seconds on each build but it's not really necessary. If you don't mind that some package lists are downloaded twice (only when `$TRAVIS_DEBIAN_EXTRA_REPOSITORY` is set anyway), I will remove it.

File: script.sh (hunk 12 above)
Comment (lamby) [reply script.sh:212]: please go ahead

File: script.sh (hunk 13 above)
Comment (lamby) [reply script.sh]: please do

**Review by lamby — CHANGES_REQUESTED** [review]

**Comment by xhaakon** [comment]
New revision is ready

File: script.sh [hunk 14]
```diff
@@ -208,14 +210,45 @@ RUN echo "deb-src ${TRAVIS_DEBIAN_MIRROR} experimental main" >> /etc/apt/sources
 EOF
 fi
 
+EXTRA_PACKAGES=""
+
+case "${TRAVIS_DEBIAN_EXTRA_REPOSITORY}" in
+	https:*)
+		EXTRA_PACKAGES="apt-transport-https"
```
Comment (lamby) [inline script.sh]: Right, but then you need to do `EXTRA_PACKAGES="${EXTRA_PACKAGES} apt-transport-https"`  :)

Comment (xhaakon) [reply script.sh]: Even though `$EXTRA_PACKAGES` is always an empty string here? I understand it's preferable to do it in case someone later puts new code that modifies `EXTRA_PACKAGES` before the case block, but I don't see why I "need to do it" - I tested the script and it works just fine. (Just being curious, I'm updating the code anyway.)

**Comment by xhaakon** [comment]
Updated. Hopefully for the last time :)

File: script.sh (hunk 14 above)
Comment (lamby) [reply script.sh]: > case someone later puts new code that modifies EXTRA_PACKAGES

Exactly that reason. We can do better than "it works" — why leave traps lying around?  :p
